File: backend/app/api/events.py
from __future__ import annotations

from typing import Optional
from fastapi import APIRouter, HTTPException, Query, status
import structlog

from app.event_engine.models import (
    CanonicalEvent,
    EventCreate,
    EventUpdate,
    EventScoreSnapshot,
    EventComparable,
    PredictionSnapshot,
)
from app.event_engine.service import event_engine_service

logger = structlog.get_logger()

router = APIRouter(prefix="/api/v1/events", tags=["Event Intelligence Engine"])


# ============================================================
# 1. Static Query Endpoints (Must precede /{event_id})
# ============================================================

@router.get("/upcoming", response_model=list[CanonicalEvent])
async def get_upcoming_events(limit: int = Query(default=50, ge=1, le=100)):
    """Retrieve upcoming scheduled, approaching, or active market events."""
    if not event_engine_service._initialized:
        await event_engine_service.initialize()
    return event_engine_service.get_upcoming_events(limit=limit)


@router.get("/today", response_model=list[CanonicalEvent])
async def get_today_events():
    """Retrieve market events scheduled or occurring today (IST)."""
    if not event_engine_service._initialized:
        await event_engine_service.initialize()
    return event_engine_service.get_today_events()


@router.get("/alerts/queue")
async def get_alerts_queue():
    """Retrieve internal alert and review queue items (§29)."""
    from app.event_engine.alert_service import event_alert_service
    return event_alert_service.get_all_alerts()


@router.post("/alerts/{alert_id}/ack")
async def acknowledge_alert(alert_id: str, user: str = Query(default="OPS_DESK")):
    """Acknowledge an alert in the internal review queue."""
    from app.event_engine.alert_service import event_alert_service
    alert = event_alert_service.acknowledge_alert(alert_id, user_name=user)
    if not alert:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Alert '{alert_id}' not found")
    return alert


@router.get("/shadow-signals")
async def get_shadow_signals():
    """Retrieve forward-testing signals operating in SHADOW_MODE (§3, §27)."""
    from app.event_engine.signal_bridge import event_signal_bridge
    return event_signal_bridge.get_shadow_records()


@router.post("/rbi/sync", response_model=list[CanonicalEvent])
async def sync_rbi_events():
    """Trigger synchronization from official Reserve Bank of India source adapter."""
    events = await event_engine_service.sync_rbi_events()
    return events


@router.post("/corporate/sync", response_model=list[CanonicalEvent])
async def sync_corporate_events():
    """Trigger synchronization from official NSE and BSE corporate source adapters."""
    return await event_engine_service.sync_corporate_events()


@router.post("/manual", response_model=CanonicalEvent, status_code=status.HTTP_201_CREATED)
async def create_manual_event(payload: EventCreate):
    """Admin/Operations endpoint for manual event entry or official fallback (§5)."""
    return await event_engine_service.create_manual_event(payload)


@router.get("/analytics/track-record")
async def get_track_record():
    """Statistical Edge Validation & Track Record Summary (§23, §24)."""
    if not event_engine_service._initialized:
        await event_engine_service.initialize()
    return event_engine_service.get_track_record()


@router.get("/sources/health")
async def get_sources_health():
    """Source adapter health telemetry, latencies, and circuit breaker status (§5, §34)."""
    return event_engine_service.get_source_health()


@router.get("/risk/overlay")
async def get_risk_overlay(underlying: str = Query(default="BANKNIFTY")):
    """Event-aware risk overlay parameters and proximity dampeners (§28)."""
    if not event_engine_service._initialized:
        await event_engine_service.initialize()
    return event_engine_service.get_risk_overlay(underlying=underlying)


@router.post("/calibrate")
async def calibrate_events():
    """Trigger empirical market impact calibration across all canonical events (§14)."""
    if not event_engine_service._initialized:
        await event_engine_service.initialize()
    return event_engine_service.calibrate_all_events()


# ============================================================
# 2. Parameterized Endpoints (/{event_id}/...)
# ============================================================

@router.get("/{event_id}", response_model=CanonicalEvent)
async def get_event_detail(event_id: str):
    """Retrieve full canonical event details, score breakdown, and impact mappings."""
    if not event_engine_service._initialized:
        await event_engine_service.initialize()
    event = event_engine_service.get_event_by_id(event_id)
    if not event:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Event '{event_id}' not found",
        )
    return event


@router.get("/{event_id}/scores", response_model=EventScoreSnapshot)
async def get_event_scores(event_id: str):
    """Retrieve 3-dimension scores (Importance, Market Impact, Opportunity) and breakdown."""
    if not event_engine_service._initialized:
        await event_engine_service.initialize()
    event = event_engine_service.get_event_by_id(event_id)
    if not event or not event.scores:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Scores for event '{event_id}' not found",
        )
    return event.scores


@router.get("/{event_id}/comparables", response_model=list[EventComparable])
async def get_event_comparables(event_id: str):
    """Retrieve historical comparables adhering to strict anti-lookahead cutoffs."""
    if not event_engine_service._initialized:
        await event_engine_service.initialize()
    event = event_engine_service.get_event_by_id(event_id)
    if not event:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Event '{event_id}' not found",
        )
    return event.comparables


@router.get("/{event_id}/prediction", response_model=PredictionSnapshot)
async def get_event_prediction(event_id: str):
    """Retrieve immutable prediction snapshot with data cutoff timestamps."""
    if not event_engine_service._initialized:
        await event_engine_service.initialize()
    event = event_engine_service.get_event_by_id(event_id)
    if not event or not event.latest_prediction:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Prediction snapshot for event '{event_id}' not found",
        )
    return event.latest_prediction


@router.get("/{event_id}/live-opportunity")
async def get_live_opportunity(event_id: str):
    """Evaluate real-time Trading Opportunity Score & live options context (§15, §16, §18)."""
    if not event_engine_service._initialized:
        await event_engine_service.initialize()
    try:
        return await event_engine_service.get_live_opportunity(event_id)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


@router.get("/{event_id}/outcome")
async def get_event_outcome(event_id: str):
    """Retrieve post-event measured reaction and realized outcome (§24)."""
    from app.event_engine.outcome_recorder import event_outcome_recorder
    outcome = event_outcome_recorder.get_outcome(event_id)
    if not outcome:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Outcome for event '{event_id}' has not yet been settled",
        )
    return outcome

File: frontend/src/lib/api/events.ts
import type { ApiCore } from './client';

export function createEventsApi(core: ApiCore) {
  return {
    async getEventRiskOverlay(underlying: string = 'BANKNIFTY') {
    return core.request<import('../event-types').EventRiskParameters>(`/api/v1/events/risk/overlay?underlying=${encodeURIComponent(underlying)}`);
  },

    async getTodayEvents() {
    return core.request<import('../event-types').CanonicalEvent[]>(`/api/v1/events/today`);
  },

    async getUpcomingEvents(limit: number = 50) {
    return core.request<import('../event-types').CanonicalEvent[]>(`/api/v1/events/upcoming?limit=${limit}`);
  },

    async getEventDetail(eventId: string) {
      return core.request<import('../event-types').CanonicalEvent>(
        `/api/v1/events/${encodeURIComponent(eventId)}`,
      );
    },

    async getEventScores(eventId: string) {
      return core.request<import('../event-types').EventScoreSnapshot>(
        `/api/v1/events/${encodeURIComponent(eventId)}/scores`,
      );
    },

    async getEventComparables(eventId: string) {
      return core.request<import('../event-types').EventComparable[]>(
        `/api/v1/events/${encodeURIComponent(eventId)}/comparables`,
      );
    },

    async getEventPrediction(eventId: string) {
      return core.request<import('../event-types').PredictionSnapshot>(
        `/api/v1/events/${encodeURIComponent(eventId)}/prediction`,
      );
    },

    async getEventLiveOpportunity(eventId: string) {
      return core.request<import('../event-types').LiveOpportunityResponse>(
        `/api/v1/events/${encodeURIComponent(eventId)}/live-opportunity`,
      );
    },

    async getEventOutcome(eventId: string) {
      return core.request<import('../event-types').EventOutcome>(
        `/api/v1/events/${encodeURIComponent(eventId)}/outcome`,
      );
    },

    async getEventAlertsQueue() {
      return core.request<import('../event-types').EventAlert[]>(`/api/v1/events/alerts/queue`);
    },

    async ackEventAlert(alertId: string, user = 'OPS_DESK') {
      return core.request<import('../event-types').EventAlert>(
        `/api/v1/events/alerts/${encodeURIComponent(alertId)}/ack?user=${encodeURIComponent(user)}`,
        { method: 'POST' },
      );
    },

    async getEventShadowSignals() {
      return core.request<import('../event-types').ShadowSignalRecord[]>(`/api/v1/events/shadow-signals`);
    },

    async syncRbiEvents() {
      return core.request<import('../event-types').CanonicalEvent[]>(`/api/v1/events/rbi/sync`, {
        method: 'POST',
      });
    },

    async syncCorporateEvents() {
      return core.request<import('../event-types').CanonicalEvent[]>(`/api/v1/events/corporate/sync`, {
        method: 'POST',
      });
    },

    async createManualEvent(payload: Record<string, unknown>) {
      return core.request<import('../event-types').CanonicalEvent>(`/api/v1/events/manual`, {
        method: 'POST',
        body: JSON.stringify(payload),
      });
    },

    async getEventTrackRecord() {
      return core.request<import('../event-types').EventTrackRecord>(
        `/api/v1/events/analytics/track-record`,
      );
    },

    async getEventSourcesHealth() {
      return core.request<import('../event-types').SourceHealthTelemetry>(`/api/v1/events/sources/health`);
    },
  };
}

export type EventsApi = ReturnType<typeof createEventsApi>;
